### PR TITLE
Allow the `integration_sesion` to be set early on ActionDispatch::Integration::Runner.

### PR DESCRIPTION
**Pull request by fables-tales**
In commit fa63448420d3385dbd043aca22dba973b45b8bb2, @tenderlove changed
the behaviour of the way `integration_session` is set up in this object.
It used to be the case that the first time it was accessed, it was
memoized with nil, however, this means that if it had already been set
it was not replaced. After that commit, it is now always set to `nil` in
the execution of `before_setup`.

In RSpec, users are able to invoke `host!` in `before(:all)` blocks,
which execute well before `before_setup` is ever invoked (which happens
in what is equivalent to a `before(:each)` block, for each test. `host!`
causes the integration session to be set up to correctly change the
host, but after fa63448420d3385dbd043aca22dba973b45b8bb2 the
`integration_session` gets overwritten, meaning that users lose their
`host!` configuration (see https://github.com/rspec/rspec-rails/issues/1662).

This commit changes the behaviour back to memoizing with `nil`, as
opposed to directly overwriting with `nil`. This causes the correct
behaviour to occur in RSpec, and unless I'm mistaken will also ensure
that users who want to modify their integration sessions early in rails
will also be able to do so.

### CODE REVIEW

**Comment by sgrif** [comment]
Do we need to be setting it back to `nil` in `after_teardown` to ensure we don't have a state leak between tests?

**Comment by samphippen** [comment]
@sgrif I think reset! fixes this case, but I'll just add the `after_teardown` anyway.

**Comment by matthewd** [comment]
It's an instance variable.. the whole object goes away after teardown.

**Comment by sgrif** [comment]
Ah ok -- I wasn't sure if minitest created a new object for every test or
not. Yeah the original code should be fine then.

**Comment by tenderlove** [comment]
IDGI. If you're using before_setup, you should be calling super first, so it doesn't matter if super sets it to nil.

Is RSpec not keeping track of the class level before(:all) data and calling before_setup with that data at the instance level?  An '||= nil' makes me do "very skeptical (͡° ͜ʖ ͡°)"

**Comment by samphippen** [comment]
@tenderlove 

> IDGI. If you're using before_setup, you should be calling super first, so it doesn't matter if super sets it to nil.

in RSpec `before(:all)` happens well before `before_setup` is invoked (which happens at roughly equivalent time to `before(:each)` hooks. This means that when the `before_setup` is invoked, if an integration session has already been set, the invocation of `before_setup` erases it.

> Is RSpec not keeping track of the class level before(:all) data and calling before_setup with that data at the instance level? An '||= nil' makes me do "very skeptical (͡° ͜ʖ ͡°)"

I was also pretty skeptical of this, but was just following the pattern that was there previously. We could totally change it to set the instance variable to nil in the constructor (constructors are pretty cool!) and then not mess with it in before_setup. WDYT?

Thanks ❤️

**Comment by tenderlove** [comment]
Can we move the ivar assignment to initialize? I really dislike the ||= nil

--
Aaron Patterson
http://tenderlovemaking.com/
I'm on an iPhone so I apologize for top posting.

**Comment by samphippen** [comment]
I'm super 👍 on that.

**Comment by samphippen** [comment]
boopity boop. pushed.

File: actionpack/test/dispatch/runner_test.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+require "abstract_unit"
+
+class RunnerTest < ActiveSupport::TestCase
+  test "runner preserves the setting of integration_session" do
+    runner = Class.new do
+      def before_setup
```
Comment (tenderlove) [inline actionpack/test/dispatch/runner_test.rb]: This should call `super`

Comment (rafaelfranca) [inline actionpack/test/dispatch/runner_test.rb]: calling `super` here will make the test fail saying that there is no super class method `before_setup`.

Comment (tenderlove) [inline actionpack/test/dispatch/runner_test.rb]: Ah, doh!

**Comment by deepj** [comment]
Hello, I'm not sure if this fixed in Rails 5.0.1. My tests fail due to this with Rails 5.0.1.rc1. If I use this patch in initializers. It works.

```ruby
# This patch is based on https://github.com/rspec/rspec-rails/issues/1662
# It solves problem with setting of a host name for integration tests (request spec in our case)

url = 'https://github.com/rails/rails/pull/26235'

warn "You may use the incompatibile patch located on #{__FILE__} with your ActionPack gem. "  \
     "Please, check possible changes with the following PR #{url}. If it is possible, " \
     'remove the patch.' if Rails.version > '5.0.0.1'

module ActionDispatch
  module Integration
    module Runner
      def initialize(*args, &blk)
        super(*args, &blk)
        @integration_session = nil
      end

      def before_setup # :nodoc:
        @app = nil
        super
      end
    end
  end
end
```

I use `host!` in the following rspec construct.

```
RSpec.configure do |config|
  config.before(:context, type: :request) do
    host! ENV['APP_HOST'] || 'http://localhost'
  end
end
```

I'm getting this issue

```
Expected response to be a redirect to <http://www.example.com/users/sign_in> but was a redirect to <http://localhost/users/sign_in>.
       Expected "http://www.example.com/users/sign_in" to be === "http://localhost/users/sign_in".
```

**Comment by deepj** [comment]
Ah, I see. This is fixed in master. Is this possible to backport to Rails 5.0-stable? /cc @rafaelfranca @sgrif

**Comment by utilum** [comment]
@deepj  : [ported](#27262)

**Comment by deepj** [comment]
@utilum Great. Thanks!

**Pull request merged**